File: app/src/main/java/com/relic/presentation/displaysub/DisplaySubContract.java
package com.relic.presentation.displaysub;

import android.arch.lifecycle.LiveData;
import android.view.View;

import com.relic.data.PostRepository;
import com.relic.data.SubRepository;
import com.relic.data.models.PostModel;
import com.relic.data.models.SubredditModel;

import java.util.List;

public class DisplaySubContract {
  static int UPVOTE = 1;
  static int DOWNVOTE = -1;
  static int RESET = 0;

  public interface ViewModel {
    void init(String subredditName, SubRepository subRepo, PostRepository postRepo);

    /**
     * Used to check if the viewmodel has already been initialized
     * @return
     */
    boolean isInitialized();

    LiveData<SubredditModel> getSubModel();

    LiveData<List<PostModel>> getPosts();

   // LiveData<Boolean> getIsSubscribed();

    String getSubName();

    void changeSortingMethod(int sortingCode, int sortScope);

    void retrieveMorePosts(boolean resetPosts);

    void updateSubStatus(boolean subscribe);

    void visitPost(String postFullname);

    void voteOnPost(String postFullname, int voteValue);
  }
}

File: app/src/main/java/com/relic/presentation/displaysub/DisplaySubContract.kt
package com.relic.presentation.displaysub

import com.relic.data.PostSource
import com.relic.data.SortScope
import com.relic.data.SortType
import com.relic.presentation.main.RelicError
import com.relic.presentation.util.MediaType

interface DisplaySubContract {
    interface ViewModel {
        fun changeSortingMethod(sortType: SortType? = null, sortScope: SortScope? = null)

        /**
         * retrieves more posts from the network
         * @param resetPosts : indicates whether the old posts should be cleared
         */
        fun retrieveMorePosts(resetPosts: Boolean)
        fun updateSubStatus(subscribe: Boolean)
    }

}

// TODO extract out of this interface
sealed class NavigationData {
    data class ToPost(
        val postFullname: String,
        val subredditName: String,
        val commentId: String? = null
    ) : NavigationData()

    data class ToPostSource(
        val source: PostSource
    ) : NavigationData()

    data class PreviewPostSource(
        val source: PostSource
    ) : NavigationData()

    // specifically for posts
    data class ToImage(
        val thumbnail: String
    ) : NavigationData()

    data class ToExternal(
        val url: String
    ) : NavigationData()

    data class ToUserPreview(
        val username: String
    ) : NavigationData()

    data class ToUser(
        val username: String
    ) : NavigationData()

    data class ToMedia(
        val mediaType: MediaType
    ) : NavigationData()

    data class ToReply(
        val parentFullname: String
    ) : NavigationData()
}

data class DisplaySubInfoData(
    var sortingMethod: SortType,
    var sortingScope: SortScope
)


object NoResults : RelicError()
